### ISSUE TITLE
how to express -inf, nodata and undetect in a tabular data package

### ISSUE THREAD

**Issue by niconoe** [enram/vpts-dp#23]
Here is an extract of data currently generated by [vph5_to_vpts](https://github.com/enram/vptstools) :

| datetime             | height | dens   | ff       | dd       | eta    | sd_vvp             | DBZH                  | dbz    | u        | v        | gap  | w        | n_dbz | n | n_all | n_dbz_all |
|----------------------|--------|--------|----------|----------|--------|--------------------|-----------------------|--------|----------|----------|------|----------|-------|---|-------|-----------|
| 2013-12-29 00:10:00Z | 0      | nodata | nodata   | nodata   | nodata | nodata             | nodata                | nodata | nodata   | nodata   | True | nodata   | 0     | 0 | 0     | 0         |
| 2013-12-29 00:10:00Z | 200    | 0.0    | undetect | undetect | 0.0    | undetect           | -7.051925182342529    | -inf   | undetect | undetect | True | undetect | 28440 | 0 | 182   | 30397     |

We have an issue with some "non-numbers" here, because:

- in #6 : we've decided that `-inf`, `undetect` and `nodata` are important nuances that should be kept when data is converted to vpts-dp
- We are currently expressing most variables using the [frictionless number type](https://specs.frictionlessdata.io/table-schema/#number), which only allows a the following special values: `NaN`, `INF` and `-INF`

It probably makes sense to map `-inf` to `-INF`, but how to express those nuances without abusing the frictionless spec is less clear for `nodata` and `undetect`. I guess other non-numer values might also be needed in future versions of the standard, so it would be good to have a generic solution for this.

Possible approaches:
- Define our own type which is a subclass of `number` that allows other specific values (but is that possible at all?)
- Express everything as a frictionless string (and document in vpts-dp that this string is either a number, either `-inf`, `nodata` or `undetect`
- Map `nodata` to `INF` and `undetect` to `NaN` (or the opposite). That technically solve the issue, but would not be very readable and could even create confusion for humans having a look at the data file.
- other?

**Comment by peterdesmet** [comment]
My thoughts: I would keep with the frictionless specs and not define our own subclass. The [ODIM bird profile specs](https://github.com/adokter/vol2bird/wiki/ODIM-bird-profile-format-specification) allow to define the `nodata (NaN)` and `undetect (9999.0)` values (defaults in parenthesis), but I would not do that either. The defaults there could be indication of what to use though. However, I'm not too big a fan of `9999.0`. 

concept | value | remark
--- | --- | ---
-inf | -inf
inf | inf
nodata | NaN
undetect | (empty value) | biggest disadvantage is that the column can then no longer be made required

@adokter any suggestions?

**Comment by adokter** [comment]
Hi noticed that the [ODIM bird profile specs](https://github.com/adokter/vol2bird/wiki/ODIM-bird-profile-format-specification) were out of date for the nodata and undetect encodings (corrected now), they have been for quite a long time `nodata -1000` and `undetect -999`, both are hard-coded in vol2bird here:
https://github.com/adokter/vol2bird/blob/8a23dfb5d3e9d96d32964fd97500c8c495f48580/lib/constants.h#L117-L122.

These negative values are very unlikely to ever interfere with true data values (reflectivities dBZ can be negative, but never this negative). Could we simply use these same integer values for the csv containing data? The nodata and undetect encodings are stored in the hdf5 profile (under `/dataset1/data1/what/nodata` and `/dataset1/data1/what/undetect`) and could even be stored as metadata of the frictionless data package.

For the parser loading csv data my suggestion would be to stay as close as possible to the bioRad conventions:
http://adriaandokter.com/bioRad/reference/summary.vp.html#conventions
so map `nodata (-1000) -> NA` and `undetect (-999) -> NaN`.

Also note as additional confusing factor that the stdout of vol2bird does not differentiate between nodata and undetect (because of limitation of the C print function) and outputs NaN for both undetect and nodata, which is something that still needs to be corrected in https://github.com/adokter/vol2bird/issues/97

So my suggestion would be

| concept      | value csv | value parsed csv |
| ----------- | ----------- |----------- |
| -inf             | -inf       | -inf
| inf              |  inf       | inf |
| nodata      |  -1000    | NA |
| undetect     |  -999    | NaN |

**Comment by peterdesmet** [comment]
Thanks @adokter for the mapping. I'm a bit more inclined to actually store `NaN` and `(empty)` in the raw csv files though, since they are already generated by a parser (https://github.com/enram/vptstools).

**Comment by adokter** [comment]
Ok, I'm a little concerned that if you flip NaN from undetect to nodata (as you suggest in https://github.com/enram/vpts-dp/issues/23#issuecomment-900108234) that might be confusing.

**Comment by niconoe** [comment]
Thanks a lot for the feedback! I'd just like to confirm/clarify two details, to make sure I fully understand the discussion:

@adokter: concerning:

> For the parser loading csv data my suggestion would be to stay as close as possible to the bioRad conventions:
http://adriaandokter.com/bioRad/reference/summary.vp.html#conventions
so map nodata (-1000) -> NA and undetect (-999) -> NaN.

(and also the "value parsed csv" column in your table): if I'm not mistaken, those suggestions about how a tool that consumes a `vpts-dp` package and its included CSV file should further transform / interpret the values from the second column, is this correct? If so, I'd be inclined  to say it's not really our concern at this step (the standard definition), and that our responsibility would be limited to properly document the available values in CSV: "you can find the special values -inf, inf, -1000 and -999, here is the meaning for each of those". Does that make sense?

@peterdesmet: concerning:

> I'm a bit more inclined to actually store NaN and (empty) in the raw csv files though, since they are already generated by a parser (https://github.com/enram/vptstools)

I'm not sure exactly what you mean with the second part of the sentence, but to me the `vph5_to_vpts` output format is currently not stabilised at all, and should be adapted to follow a proper standard (as discussed here), rather than the other way around (the idea to develop `vptstools` in parallel to the standard discussions was to raise important questions early, rather than to freeze specifications prematurely). Does that make sense?

**Comment by peterdesmet** [comment]
@niconoe @adokter  I meant that I think the CSV files should store `NA` (not -1000) for `nodata` and `NaN` (not -999) for `undetect`, as NA and NaN contain more intrinsic meaning than -1000 and -999.

In addition, I think the `missingValues` property of the table schema should contain `NA`, so empty and NA values are treated the same when reading.

**Comment by niconoe** [comment]
@peterdesmet: if I understand the [frictionless number type spec](https://specs.frictionlessdata.io/table-schema/#number) correctly, `NA` is not allowed there (only `NaN`, `Inf` and `-Inf`).

**Comment by peterdesmet** [comment]
Before values are even interpreted as a type, you can define which ones should be considered missing values in https://specs.frictionlessdata.io/table-schema/#missing-values. If you list `NA` there (in addition to the default `""`) it will be considered empty. Important to note is that there is basically no difference then between `""` and `NA`: if one of those occur and the field is `required`, it will fail validation.

So, I think we can choose how to express `nodata`: either as `""` (empty value) or `NA`. If a column is expected to have such `nodata` values, we cannot make it required.

**Comment by peterdesmet** [comment]
Picking up this discussion again.

## Inf / -Inf

store in CSV as | readr::read_csv() | frictionless
--- | --- | ---
`Inf` and `-Inf` | Interpreted correctly be default (case insensitive) | Allowed as number (case insensitive), but will expectedly fail e.g. `minimum: -10`

## Nodata

Coded as `-1000` by ODIM, interpreted as `NA` by bioRad

store in CSV as | readr::read_csv() | frictionless
--- | --- | ---
`-1000` | Not interpreted as NA by default, needs conversion | Allowed (is a number), but will expectedly fail e.g. `minimum: -10`
`NA` | Interpreted as NA by default | Allowed if [missingValues](https://specs.frictionlessdata.io/table-schema/#missing-values) includes `NA`, expectedly bypasses e.g. `minimum: -10`. Interpreted exactly the same as empty value.
empty value | Interpreted as NA by default | Allowed by default, expectedly bypasses e.g. `minimum: -10`

## Undetect

Coded as `-999` by ODIM, interpreted as `NaN` by bioRad

store in CSV as | readr::read_csv() | frictionless
--- | --- | ---
`-999` | Not interpreted as NaN by default, needs conversion | Allowed (is a number), but will expectedly fail e.g. `minimum: -10`
`NaN` | Interpreted as NaN by default (case insensitive) | Allowed by default (case insensitive), but will expectedly fail e.g. `minimum: -10`

## Recommendation

Given the above, I suggest to:

- Store `Inf` and `-Inf` as `Inf`/`INF`/`inf`/ and `-Inf`,`-INF`, `-inf`
- Store `nodata` as empty values or `NA` (but not `Na` or `na`)
- Store `undetect` as `NaN`/`NAN`/`nan`

@niconoe and @adokter is that ok for you?

**Comment by niconoe** [comment]
Sorry to not answer yet and ask one more question instead, but:

If VPTS has now evolved and is not a frictionless data package anymore, shouldn't we drop their data and requirements completely, and gain some control and flexibility in exchange? That would make this issue simpler/nonexistent (we allow the specific `-inf`, `nodata` and `undetect` values, properly document what they mean, and we're done).

I have the same doubts but about tooling for issue #14: if we're not fully embracing data packages and producing data packages (with a `datapackage.json` and so on), I don't see the point to use their part of their tooling, in addition to the checks we'll have to implement ourselves. More specifically, I see the point in theory (reusability, existing standards) but I'm afraid in practice the disadvantages (more dependencies) might be more important than the advantages.

**Comment by peterdesmet** [comment]
I understand your concerns regarding using Frictionless tooling. However, the suggestion is not only one for Frictionless. Using string values `nodata` and `nodetect` in numeric columns (rather than `NA`/`NaN` or `-1000`/`-999` for that matter) is going to be more of a hassle when reading CSV files with other tools (readr, pandas, …). I see the conversion to `NA` and `NaN` by bioRad (the most used tool for these files) as a good indication to store the value as such. As a bonus, it works well with readr, pandas, and Frictionless.

**Comment by niconoe** [comment]
Thanks, I had no idea there was some de facto standardisation about those values in CSV files!

**Comment by peterdesmet** [comment]
I have now added:

> Field values SHOULD meet the constraints as defined below, except for `NaN`, `NA` and empty values.

@adokter can you tell me:

- If you agree with my recommendation in https://github.com/enram/vpts/issues/23#issuecomment-1119576827
- If you have a preference for empty or "NA" in the CSV for nodata? NA is more verbose
- If I should clarify the undetect/nodata in the format specs?

**Comment by adokter** [comment]
Hi @peterdesmet, looks good. I prefer NA over empty, since that lines it up more closely with the bioRad convention.

The -1000 and -900 values are not part of the ODIM specification, but they are introduced by bioRad as a hard-coded value, see  https://github.com/adokter/vol2bird/blob/61ad71ae02bbd1ad1a8bc75792f757ea9d281ce0/lib/constants.h#L117-L121

The ODIM standard requires definition of `undetect` and `nodata` attributes to decode the data. vol2bird follows that standard, and adds these two attributes to the hdf5 profiles, with assigned values `-999` and `-1000`
